### PR TITLE
Fix BlockPublisher handling of failed batches, Such that if a Candida…

### PR DESCRIPTION
**Pull request by cintel**
…teBlock contains only Batches that fail the failed batches will be discarded and will not be retried.

Signed-off-by: Cian Montgomery <cian.montgomery@intel.com>

### CODE REVIEW

**Comment by cintel** [comment]
@boydjohnson  can you test this in accordance with the direction you left here: https://jira.hyperledger.org/browse/STL-158

I applied the patch and ran the test but the description was unclear what I was looking for either as a positive or negative result.

**Comment by boydjohnson** [comment]
Yeah, I'll test it, but I won't have time until tomorrow morning.

**Comment by nick-drozd** [comment]
No tests need to be modified to see the difference. Here's what happens when an invalid batch is sent in `intkey-python-smoke` before the change:

```
integration_test_1  | INFO:test_intkey_smoke:Posting batch
rest_api_1          | [13:21:23.730 INFO     rest_api] Request  35aeb6: "POST /batches" from 172.18.0.6
rest_api_1          | [13:21:23.746 INFO     rest_api] Response 35aeb6: 202 status, 181B size, in 0.016s
tp_intkey_python_1  | [13:21:23.750 DEBUG    core] received message of type: TP_PROCESS_REQUEST
tp_intkey_python_1  | [13:21:23.760 DEBUG    handler] Incrementing "cow" by 1
tp_intkey_python_1  | [13:21:23.760 WARNING  core] Invalid Transaction Verb is "inc" but name "cow" not in state
tp_intkey_python_1  | [13:21:23.765 DEBUG    core] received message of type: TP_PROCESS_REQUEST
tp_intkey_python_1  | [13:21:23.770 DEBUG    handler] Incrementing "pig" by 2
tp_intkey_python_1  | [13:21:23.773 WARNING  core] Invalid Transaction Verb is "inc" but name "pig" not in state
tp_intkey_python_1  | [13:21:23.791 DEBUG    core] received message of type: TP_PROCESS_REQUEST
tp_intkey_python_1  | [13:21:23.799 DEBUG    handler] Decrementing "sheep" by 3
tp_intkey_python_1  | [13:21:23.800 WARNING  core] Invalid Transaction Verb is "dec" but name "sheep" not in state
tp_intkey_python_1  | [13:21:23.807 DEBUG    core] received message of type: TP_PROCESS_REQUEST
tp_intkey_python_1  | [13:21:23.813 DEBUG    handler] Incrementing "goat" by 5
tp_intkey_python_1  | [13:21:23.813 WARNING  core] Invalid Transaction Verb is "inc" but name "goat" not in state
tp_intkey_python_1  | [13:21:23.824 DEBUG    core] received message of type: TP_PROCESS_REQUEST
tp_intkey_python_1  | [13:21:23.842 DEBUG    handler] Decrementing "horse" by 8
tp_intkey_python_1  | [13:21:23.843 WARNING  core] Invalid Transaction Verb is "dec" but name "horse" not in state
tp_intkey_python_1  | [13:21:23.949 DEBUG    core] received message of type: TP_PROCESS_REQUEST
tp_intkey_python_1  | [13:21:23.959 DEBUG    handler] Incrementing "cow" by 1
tp_intkey_python_1  | [13:21:23.959 WARNING  core] Invalid Transaction Verb is "inc" but name "cow" not in state
tp_intkey_python_1  | [13:21:23.964 DEBUG    core] received message of type: TP_PROCESS_REQUEST
tp_intkey_python_1  | [13:21:23.968 DEBUG    handler] Incrementing "pig" by 2
tp_intkey_python_1  | [13:21:23.968 WARNING  core] Invalid Transaction Verb is "inc" but name "pig" not in state
tp_intkey_python_1  | [13:21:23.972 DEBUG    core] received message of type: TP_PROCESS_REQUEST
tp_intkey_python_1  | [13:21:23.976 DEBUG    handler] Decrementing "sheep" by 3
tp_intkey_python_1  | [13:21:23.976 WARNING  core] Invalid Transaction Verb is "dec" but name "sheep" not in state
tp_intkey_python_1  | [13:21:23.981 DEBUG    core] received message of type: TP_PROCESS_REQUEST
tp_intkey_python_1  | [13:21:23.984 DEBUG    handler] Incrementing "goat" by 5
tp_intkey_python_1  | [13:21:23.984 WARNING  core] Invalid Transaction Verb is "inc" but name "goat" not in state
tp_intkey_python_1  | [13:21:23.994 DEBUG    core] received message of type: TP_PROCESS_REQUEST
tp_intkey_python_1  | [13:21:23.998 DEBUG    handler] Decrementing "horse" by 8
tp_intkey_python_1  | [13:21:23.999 WARNING  core] Invalid Transaction Verb is "dec" but name "horse" not in state
tp_intkey_python_1  | [13:21:24.106 DEBUG    core] received message of type: TP_PROCESS_REQUEST
tp_intkey_python_1  | [13:21:24.110 DEBUG    handler] Incrementing "cow" by 1
tp_intkey_python_1  | [13:21:24.111 WARNING  core] Invalid Transaction Verb is "inc" but name "cow" not in state
tp_intkey_python_1  | [13:21:24.116 DEBUG    core] received message of type: TP_PROCESS_REQUEST
tp_intkey_python_1  | [13:21:24.120 DEBUG    handler] Incrementing "pig" by 2
tp_intkey_python_1  | [13:21:24.120 WARNING  core] Invalid Transaction Verb is "inc" but name "pig" not in state
tp_intkey_python_1  | [13:21:24.125 DEBUG    core] received message of type: TP_PROCESS_REQUEST
tp_intkey_python_1  | [13:21:24.129 DEBUG    handler] Decrementing "sheep" by 3
tp_intkey_python_1  | [13:21:24.129 WARNING  core] Invalid Transaction Verb is "dec" but name "sheep" not in state
tp_intkey_python_1  | [13:21:24.134 DEBUG    core] received message of type: TP_PROCESS_REQUEST
tp_intkey_python_1  | [13:21:24.138 DEBUG    handler] Incrementing "goat" by 5
tp_intkey_python_1  | [13:21:24.138 WARNING  core] Invalid Transaction Verb is "inc" but name "goat" not in state
tp_intkey_python_1  | [13:21:24.143 DEBUG    core] received message of type: TP_PROCESS_REQUEST
tp_intkey_python_1  | [13:21:24.150 DEBUG    handler] Decrementing "horse" by 8
tp_intkey_python_1  | [13:21:24.151 WARNING  core] Invalid Transaction Verb is "dec" but name "horse" not in state
tp_intkey_python_1  | [13:21:24.260 DEBUG    core] received message of type: TP_PROCESS_REQUEST
tp_intkey_python_1  | [13:21:24.264 DEBUG    handler] Incrementing "cow" by 1
tp_intkey_python_1  | [13:21:24.264 WARNING  core] Invalid Transaction Verb is "inc" but name "cow" not in state
tp_intkey_python_1  | [13:21:24.269 DEBUG    core] received message of type: TP_PROCESS_REQUEST
tp_intkey_python_1  | [13:21:24.273 DEBUG    handler] Incrementing "pig" by 2
tp_intkey_python_1  | [13:21:24.273 WARNING  core] Invalid Transaction Verb is "inc" but name "pig" not in state
tp_intkey_python_1  | [13:21:24.278 DEBUG    core] received message of type: TP_PROCESS_REQUEST
tp_intkey_python_1  | [13:21:24.281 DEBUG    handler] Decrementing "sheep" by 3
tp_intkey_python_1  | [13:21:24.282 WARNING  core] Invalid Transaction Verb is "dec" but name "sheep" not in state
tp_intkey_python_1  | [13:21:24.286 DEBUG    core] received message of type: TP_PROCESS_REQUEST
tp_intkey_python_1  | [13:21:24.290 DEBUG    handler] Incrementing "goat" by 5
tp_intkey_python_1  | [13:21:24.290 WARNING  core] Invalid Transaction Verb is "inc" but name "goat" not in state
tp_intkey_python_1  | [13:21:24.295 DEBUG    core] received message of type: TP_PROCESS_REQUEST
tp_intkey_python_1  | [13:21:24.303 DEBUG    handler] Decrementing "horse" by 8
tp_intkey_python_1  | [13:21:24.303 WARNING  core] Invalid Transaction Verb is "dec" but name "horse" not in state
tp_intkey_python_1  | [13:21:24.411 DEBUG    core] received message of type: TP_PROCESS_REQUEST
tp_intkey_python_1  | [13:21:24.415 DEBUG    handler] Incrementing "cow" by 1
tp_intkey_python_1  | [13:21:24.416 WARNING  core] Invalid Transaction Verb is "inc" but name "cow" not in state
tp_intkey_python_1  | [13:21:24.421 DEBUG    core] received message of type: TP_PROCESS_REQUEST
tp_intkey_python_1  | [13:21:24.426 DEBUG    handler] Incrementing "pig" by 2
tp_intkey_python_1  | [13:21:24.426 WARNING  core] Invalid Transaction Verb is "inc" but name "pig" not in state
tp_intkey_python_1  | [13:21:24.431 DEBUG    core] received message of type: TP_PROCESS_REQUEST
tp_intkey_python_1  | [13:21:24.435 DEBUG    handler] Decrementing "sheep" by 3
tp_intkey_python_1  | [13:21:24.435 WARNING  core] Invalid Transaction Verb is "dec" but name "sheep" not in state
tp_intkey_python_1  | [13:21:24.442 DEBUG    core] received message of type: TP_PROCESS_REQUEST
tp_intkey_python_1  | [13:21:24.448 DEBUG    handler] Incrementing "goat" by 5
tp_intkey_python_1  | [13:21:24.448 WARNING  core] Invalid Transaction Verb is "inc" but name "goat" not in state
tp_intkey_python_1  | [13:21:24.455 DEBUG    core] received message of type: TP_PROCESS_REQUEST
tp_intkey_python_1  | [13:21:24.463 DEBUG    handler] Decrementing "horse" by 8
tp_intkey_python_1  | [13:21:24.463 WARNING  core] Invalid Transaction Verb is "dec" but name "horse" not in state
tp_intkey_python_1  | [13:21:24.571 DEBUG    core] received message of type: TP_PROCESS_REQUEST
tp_intkey_python_1  | [13:21:24.575 DEBUG    handler] Incrementing "cow" by 1
tp_intkey_python_1  | [13:21:24.575 WARNING  core] Invalid Transaction Verb is "inc" but name "cow" not in state
tp_intkey_python_1  | [13:21:24.580 DEBUG    core] received message of type: TP_PROCESS_REQUEST
tp_intkey_python_1  | [13:21:24.584 DEBUG    handler] Incrementing "pig" by 2
tp_intkey_python_1  | [13:21:24.585 WARNING  core] Invalid Transaction Verb is "inc" but name "pig" not in state
tp_intkey_python_1  | [13:21:24.589 DEBUG    core] received message of type: TP_PROCESS_REQUEST
tp_intkey_python_1  | [13:21:24.593 DEBUG    handler] Decrementing "sheep" by 3
tp_intkey_python_1  | [13:21:24.594 WARNING  core] Invalid Transaction Verb is "dec" but name "sheep" not in state
tp_intkey_python_1  | [13:21:24.598 DEBUG    core] received message of type: TP_PROCESS_REQUEST
tp_intkey_python_1  | [13:21:24.602 DEBUG    handler] Incrementing "goat" by 5
tp_intkey_python_1  | [13:21:24.602 WARNING  core] Invalid Transaction Verb is "inc" but name "goat" not in state
tp_intkey_python_1  | [13:21:24.607 DEBUG    core] received message of type: TP_PROCESS_REQUEST
tp_intkey_python_1  | [13:21:24.614 DEBUG    handler] Decrementing "horse" by 8
tp_intkey_python_1  | [13:21:24.615 WARNING  core] Invalid Transaction Verb is "dec" but name "horse" not in state
tp_intkey_python_1  | [13:21:24.723 DEBUG    core] received message of type: TP_PROCESS_REQUEST
tp_intkey_python_1  | [13:21:24.728 DEBUG    handler] Incrementing "cow" by 1
tp_intkey_python_1  | [13:21:24.728 WARNING  core] Invalid Transaction Verb is "inc" but name "cow" not in state
tp_intkey_python_1  | [13:21:24.733 DEBUG    core] received message of type: TP_PROCESS_REQUEST
tp_intkey_python_1  | [13:21:24.737 DEBUG    handler] Incrementing "pig" by 2
tp_intkey_python_1  | [13:21:24.737 WARNING  core] Invalid Transaction Verb is "inc" but name "pig" not in state
tp_intkey_python_1  | [13:21:24.741 DEBUG    core] received message of type: TP_PROCESS_REQUEST
tp_intkey_python_1  | [13:21:24.745 DEBUG    handler] Decrementing "sheep" by 3
tp_intkey_python_1  | [13:21:24.745 WARNING  core] Invalid Transaction Verb is "dec" but name "sheep" not in state
rest_api_1          | [13:21:24.750 INFO     rest_api] Request  35d694: "GET /state" from 172.18.0.6
tp_intkey_python_1  | [13:21:24.751 DEBUG    core] received message of type: TP_PROCESS_REQUEST
tp_intkey_python_1  | [13:21:24.756 DEBUG    handler] Incrementing "goat" by 5
tp_intkey_python_1  | [13:21:24.756 WARNING  core] Invalid Transaction Verb is "inc" but name "goat" not in state
tp_intkey_python_1  | [13:21:24.768 DEBUG    core] received message of type: TP_PROCESS_REQUEST
tp_intkey_python_1  | [13:21:24.774 DEBUG    handler] Decrementing "horse" by 8
tp_intkey_python_1  | [13:21:24.774 WARNING  core] Invalid Transaction Verb is "dec" but name "horse" not in state
rest_api_1          | [13:21:24.806 INFO     rest_api] Response 35d694: 200 status, 1069B size, in 0.055s
integration_test_1  | INFO:test_intkey_smoke:Current state: {'wren': 442, 'thrush': 327, 'jay': 535, 'lark': 416, 'finch': 643}
```

Here's what happens with this PR:

```
integration_test_1  | INFO:test_intkey_smoke:Posting batch
rest_api_1          | [13:26:11.732 INFO     rest_api] Request  61a0cf: "POST /batches" from 172.18.0.6
rest_api_1          | [13:26:11.746 INFO     rest_api] Response 61a0cf: 202 status, 181B size, in 0.014s
tp_intkey_python_1  | [13:26:11.750 DEBUG    core] received message of type: TP_PROCESS_REQUEST
tp_intkey_python_1  | [13:26:11.756 DEBUG    handler] Incrementing "cow" by 1
tp_intkey_python_1  | [13:26:11.758 WARNING  core] Invalid Transaction Verb is "inc" but name "cow" not in state
tp_intkey_python_1  | [13:26:11.776 DEBUG    core] received message of type: TP_PROCESS_REQUEST
tp_intkey_python_1  | [13:26:11.789 DEBUG    handler] Incrementing "pig" by 2
tp_intkey_python_1  | [13:26:11.790 WARNING  core] Invalid Transaction Verb is "inc" but name "pig" not in state
tp_intkey_python_1  | [13:26:11.795 DEBUG    core] received message of type: TP_PROCESS_REQUEST
tp_intkey_python_1  | [13:26:11.805 DEBUG    handler] Decrementing "sheep" by 3
tp_intkey_python_1  | [13:26:11.806 WARNING  core] Invalid Transaction Verb is "dec" but name "sheep" not in state
tp_intkey_python_1  | [13:26:11.814 DEBUG    core] received message of type: TP_PROCESS_REQUEST
tp_intkey_python_1  | [13:26:11.823 DEBUG    handler] Incrementing "goat" by 5
tp_intkey_python_1  | [13:26:11.825 WARNING  core] Invalid Transaction Verb is "inc" but name "goat" not in state
tp_intkey_python_1  | [13:26:11.830 DEBUG    core] received message of type: TP_PROCESS_REQUEST
tp_intkey_python_1  | [13:26:11.854 DEBUG    handler] Decrementing "horse" by 8
tp_intkey_python_1  | [13:26:11.857 WARNING  core] Invalid Transaction Verb is "dec" but name "horse" not in state
rest_api_1          | [13:26:12.753 INFO     rest_api] Request  61c8ae: "GET /state" from 172.18.0.6
rest_api_1          | [13:26:12.774 INFO     rest_api] Response 61c8ae: 200 status, 1069B size, in 0.021s
integration_test_1  | INFO:test_intkey_smoke:Current state: {'wren': 442, 'jay': 535, 'lark': 416, 'finch': 643, 'thrush': 327}
```

So it seems to work.

Some extra regression test cases should probably be added to the publisher unit tests though.

File: validator/sawtooth_validator/journal/publisher.py [hunk 1]
```diff
@@ -194,6 +194,9 @@ def finalize_block(self, identity_signing_key, pending_batches):
         :param pending_batches: list to receive any batches that were
         submitted to add to the block but were not validated before this
         call.
+        :return: None if the block failed to finalize or the Block that was
```
Comment (delventhalz) [inline validator/sawtooth_validator/journal/publisher.py]: Unclear if the _"or"_ refers to under what conditions `None` will be returned, or if it refers to the possible return values. Suggested rewording:

_The generated Block, or None if Block failed to finalize._

File: validator/sawtooth_validator/journal/publisher.py [hunk 2]
```diff
@@ -194,6 +194,9 @@ def finalize_block(self, identity_signing_key, pending_batches):
...
+        generated. In both case the pending_batches will contain the list
```
Comment (delventhalz) [inline validator/sawtooth_validator/journal/publisher.py]: case -> cases

File: validator/sawtooth_validator/journal/publisher.py [hunk 3]
```diff
@@ -204,7 +207,22 @@ def finalize_block(self, identity_signing_key, pending_batches):
         committed_txn_cache = TransactionCache(self._block_store)
 
         builder = self._block_builder
+        bad_batches = []  # the list of batches that failed processing, used
```
Comment (delventhalz) [inline validator/sawtooth_validator/journal/publisher.py]: > _", used"_

??

File: validator/sawtooth_validator/journal/publisher.py [hunk 4]
```diff
@@ -204,7 +207,22 @@ def finalize_block(self, identity_signing_key, pending_batches):
...
         state_hash = None
+
+        # Walk the pending batch list:
+        # - find the state hash for the block, the block state_hash is
+        # is randomly placed on one of the transactions, so must interogate
+        # every batch to find it. If it is on a batch that failed processing
+        # then this block will be abandoned.
+        # - build three lists of batches.
```
Comment (delventhalz) [inline validator/sawtooth_validator/journal/publisher.py]: three lists of batches. -> three lists of batches:

File: validator/sawtooth_validator/journal/publisher.py [hunk 5]
```diff
@@ -204,7 +207,22 @@ def finalize_block(self, identity_signing_key, pending_batches):
...
+        # 1) a lists of all valid transactions that will be included in the
+        # block, these are added to the BlockBuilder to include in the Block
+        # 2) all batches that were not executed, these are to be returned
+        # in the pending_batches list
+        # 3) all batches that failed processing. These will be discarded. This
+        # list is needed in some case when the block is abandoned to make
+        # sure they do not remain in the pending_batches list.
```
Comment (delventhalz) [inline validator/sawtooth_validator/journal/publisher.py]: Indentation would help clarify this comment. It is currently very difficult to read.

File: validator/sawtooth_validator/journal/publisher.py [hunk 6]
```diff
@@ -227,24 +245,37 @@ def finalize_block(self, identity_signing_key, pending_batches):
                     LOGGER.debug("Abandoning block %s:" +
                                  "root state hash has invalid txn applied",
                                  builder)
+                    # update the pending batch list to be all the
+                    # batches that passed validation to this point and
+                    # none of the ones that failed. It is possible that
+                    # this batch caused a future batch to fail so
+                    # we leave all of the batches that failed passed this
```
Comment (delventhalz) [inline validator/sawtooth_validator/journal/publisher.py]: > _"that failed passed this"_

That failed _past_ this? If so, _"failed after"_ may be more clear.

File: validator/sawtooth_validator/journal/publisher.py [hunk 7]
```diff
@@ -227,24 +245,37 @@ def finalize_block(self, identity_signing_key, pending_batches):
...
                 if result.state_hash is not None:
                     state_hash = result.state_hash
             else:
+                bad_batches.append(batch)
                 LOGGER.debug("Batch %s invalid, not added to block.",
                              batch.header_signature)
 
-        if state_hash is None:
+        if state_hash is None or not builder.batches:
             LOGGER.debug("Abandoning block %s no batches added", builder)
```
Comment (delventhalz) [inline validator/sawtooth_validator/journal/publisher.py]: Abandoning block %s -> Abandoning block %s:

Add a colon or some other punctuation to clarify that _"no batches added"_ is a separate clause, the reason for the abandonment.

File: validator/sawtooth_validator/journal/publisher.py [hunk 8]
```diff
@@ -463,21 +494,22 @@ def on_check_publish_block(self, force=False):
                         self._identity_signing_key,
                         pending_batches)
                     self._candidate_block = None
+                    # In every case we get a pending batch list back.
```
Comment (delventhalz) [inline validator/sawtooth_validator/journal/publisher.py]: ~# In every case we get a pending batch list back.~

Unnecessary.

File: validator/sawtooth_validator/journal/publisher.py [hunk 9]
```diff
@@ -463,21 +494,22 @@ def on_check_publish_block(self, force=False):
...
+                    # Update the _pending_batches to reflect what we learned.
+                    last_batch_index = self._pending_batches.index(last_batch)
+                    additional_batches =\
+                        self._pending_batches[last_batch_index + 1:]
+                    # the pending batch list is now the list we have returned
+                    # from the candidate block plus all batche that were not
```
Comment (delventhalz) [inline validator/sawtooth_validator/journal/publisher.py]: batche -> batches

Also, I'm not sure this comment is necessary at all if you rename `additional_batches` to `unsent_batches` or something similar.

File: validator/sawtooth_validator/journal/publisher.py [hunk 10]
```diff
@@ -227,24 +245,37 @@ def finalize_block(self, identity_signing_key, pending_batches):
...
                                  "root state hash has invalid txn applied",
                                  builder)
+                    # Update the pending batch list to be all the
+                    # batches that passed validation to this point and
+                    # none of the ones that failed. It is possible that
+                    # this batch caused a future batch to fail so
+                    # we leave all of the batches that failed after this
+                    # one in the list.
+                    bad_batches.append(batch)
+                    pending_batches = [x for x in self._pending_batches
```
Comment (peterschwarz) [inline validator/sawtooth_validator/journal/publisher.py]: This does not get put into the target array, it just overwrites the reference.  This use will lose pending batches.

File: validator/sawtooth_validator/journal/publisher.py [hunk 11]
```diff
@@ -227,24 +245,37 @@ def finalize_block(self, identity_signing_key, pending_batches):
...
+        if state_hash is None or not builder.batches:
+            LOGGER.debug("Abandoning block %s: no batches added", builder)
             return None
 
         if not self._consensus.finalize_block(builder.block_header):
             LOGGER.debug("Abandoning block %s, consensus failed to finalize "
                          "it", builder)
-            return False
+            # return all valid batches to the pending_batches list
+            pending_batches = [x for x in self._pending_batches
```
Comment (peterschwarz) [inline validator/sawtooth_validator/journal/publisher.py]: same as above

File: validator/sawtooth_validator/journal/publisher.py (hunk 10 above)
Comment (peterschwarz) [inline validator/sawtooth_validator/journal/publisher.py]: Need to use something like

```
    pending_batches.clear()
    pending_batches.extend([x for x in source if x < 10])
```

File: validator/sawtooth_validator/journal/publisher.py [hunk 12]
```diff
@@ -204,7 +207,22 @@ def finalize_block(self, identity_signing_key, pending_batches):
         committed_txn_cache = TransactionCache(self._block_store)
 
         builder = self._block_builder
+        bad_batches = []  # the list of batches that failed processing
         state_hash = None
+
+        # Walk the pending batch list:
+        # - find the state hash for the block, the block state_hash is
+        # is randomly placed on one of the transactions, so must interogate
```
Comment (vaporos) [inline validator/sawtooth_validator/journal/publisher.py]: Is it actually random?  Seems like this should be deterministic.

**Pull request merged**